File: import-articles.py
#!/usr/bin/env python3
"""
Import blog articles from Excel file to database
"""
import openpyxl
import mysql.connector
import os
import re
from datetime import datetime
from html.parser import HTMLParser

class HTMLTextExtractor(HTMLParser):
    """Extract text content from HTML"""
    def __init__(self):
        super().__init__()
        self.text = []
    
    def handle_data(self, data):
        self.text.append(data)
    
    def get_text(self):
        return ''.join(self.text).strip()

def extract_text_from_html(html_content):
    """Extract plain text from HTML"""
    parser = HTMLTextExtractor()
    parser.feed(str(html_content))
    return parser.get_text()

def clean_html_content(content):
    """Clean HTML content by removing DOCTYPE and html/head/body tags"""
    if not content:
        return ""
    
    content = str(content)
    
    # Remove DOCTYPE
    content = re.sub(r'<!DOCTYPE[^>]*>', '', content, flags=re.IGNORECASE)
    
    # Extract body content if exists
    body_match = re.search(r'<body[^>]*>(.*?)</body>', content, re.DOTALL | re.IGNORECASE)
    if body_match:
        content = body_match.group(1)
    else:
        # Remove html, head, body tags
        content = re.sub(r'</?html[^>]*>', '', content, flags=re.IGNORECASE)
        content = re.sub(r'<head>.*?</head>', '', content, flags=re.DOTALL | re.IGNORECASE)
        content = re.sub(r'</?body[^>]*>', '', content, flags=re.IGNORECASE)
    
    return content.strip()

# Database connection
db_url = os.getenv('DATABASE_URL')
if not db_url:
    print("ERROR: DATABASE_URL environment variable not set")
    exit(1)

# Parse DATABASE_URL (format: mysql://user:pass@host:port/dbname)
match = re.match(r'mysql://([^:]+):([^@]+)@([^:]+):(\d+)/(.+)', db_url)
if not match:
    print("ERROR: Invalid DATABASE_URL format")
    exit(1)

user, password, host, port, database = match.groups()

print(f"Connecting to database: {host}:{port}/{database}")

conn = mysql.connector.connect(
    host=host,
    port=int(port),
    user=user,
    password=password,
    database=database
)
cursor = conn.cursor()

# Load Excel file
print("\nLoading Excel file...")
wb = openpyxl.load_workbook('/home/ubuntu/upload/car_insurance_articles_100_images.xlsx')
sheet = wb.active

print(f"Found {sheet.max_row - 1} articles to import\n")

# Import articles
imported = 0
skipped = 0
errors = 0

for row_num in range(2, sheet.max_row + 1):
    try:
        row = list(sheet[row_num])
        
        title = row[0].value
        slug = row[1].value
        content = clean_html_content(row[2].value)
        excerpt_raw = row[3].value
        category = row[4].value
        tags = row[5].value
        meta_title = row[6].value
        meta_description_raw = row[7].value
        image_filename = row[8].value
        
        # Extract text from HTML for excerpt and meta description
        excerpt = extract_text_from_html(excerpt_raw) if excerpt_raw else ""
        meta_description = extract_text_from_html(meta_description_raw) if meta_description_raw else ""
        
        # Truncate if too long
        if len(excerpt) > 500:
            excerpt = excerpt[:497] + "..."
        if len(meta_description) > 160:
            meta_description = meta_description[:157] + "..."
        if len(meta_title) > 60:
            meta_title = meta_title[:60]
        
        # Build image URL
        cover_image = f"/images/blog/{image_filename}" if image_filename else ""
        
        # Check if article already exists
        cursor.execute("SELECT id FROM blog_posts WHERE slug = %s", (slug,))
        if cursor.fetchone():
            print(f"⏭️  Skipping (exists): {title[:50]}...")
            skipped += 1
            continue
        
        # Insert article
        published_at = datetime.now()
        
        cursor.execute("""
            INSERT INTO blog_posts 
            (title, slug, content, excerpt, cover_image, category, tags, 
             meta_title, meta_description, status, published_at, created_at, updated_at)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        """, (
            title, slug, content, excerpt, cover_image, category, tags,
            meta_title, meta_description, 'published', published_at, 
            published_at, published_at
        ))
        
        conn.commit()
        print(f"✅ Imported: {title[:60]}...")
        imported += 1
        
    except Exception as e:
        print(f"❌ Error importing row {row_num}: {str(e)[:100]}")
        errors += 1
        conn.rollback()

cursor.close()
conn.close()

print(f"\n{'='*60}")
print(f"Import completed!")
print(f"✅ Imported: {imported}")
print(f"⏭️  Skipped: {skipped}")
print(f"❌ Errors: {errors}")
print(f"{'='*60}")

File: import-articles.mjs
#!/usr/bin/env node
/**
 * Import blog articles from Excel file to database
 */
import { readFileSync } from 'fs';
import { read, utils } from 'xlsx';
import { drizzle } from 'drizzle-orm/mysql2';
import mysql from 'mysql2/promise';
import { blogPosts } from './drizzle/schema.js';
import { eq } from 'drizzle-orm';

// Load environment variables
const DATABASE_URL = process.env.DATABASE_URL;
if (!DATABASE_URL) {
  console.error('ERROR: DATABASE_URL environment variable not set');
  process.exit(1);
}

// Helper functions
function cleanHtmlContent(content) {
  if (!content) return '';
  
  let cleaned = String(content);
  
  // Remove DOCTYPE
  cleaned = cleaned.replace(/<!DOCTYPE[^>]*>/gi, '');
  
  // Extract body content if exists
  const bodyMatch = cleaned.match(/<body[^>]*>(.*?)<\/body>/is);
  if (bodyMatch) {
    cleaned = bodyMatch[1];
  } else {
    // Remove html, head, body tags
    cleaned = cleaned.replace(/<\/?html[^>]*>/gi, '');
    cleaned = cleaned.replace(/<head>.*?<\/head>/gis, '');
    cleaned = cleaned.replace(/<\/?body[^>]*>/gi, '');
  }
  
  return cleaned.trim();
}

function extractTextFromHtml(html) {
  if (!html) return '';
  return String(html).replace(/<[^>]*>/g, '').trim();
}

function truncate(text, maxLength) {
  if (!text || text.length <= maxLength) return text;
  return text.substring(0, maxLength - 3) + '...';
}

// Main import function
async function importArticles() {
  console.log('Loading Excel file...\n');
  
  // Read Excel file
  const workbook = read(readFileSync('/home/ubuntu/upload/car_insurance_articles_100_images.xlsx'));
  const sheetName = workbook.SheetNames[0];
  const worksheet = workbook.Sheets[sheetName];
  const data = utils.sheet_to_json(worksheet);
  
  console.log(`Found ${data.length} articles to import\n`);
  
  // Connect to database
  const connection = await mysql.createConnection(DATABASE_URL);
  const db = drizzle(connection);
  
  let imported = 0;
  let skipped = 0;
  let errors = 0;
  
  for (const row of data) {
    try {
      const title = row['Title'];
      const slug = row['Slug'];
      const content = cleanHtmlContent(row['Content']);
      const excerptRaw = row['Excerpt'];
      const category = row['Category'];
      const tags = row['Tags'];
      const metaTitle = truncate(row['Meta Title'], 60);
      const metaDescriptionRaw = row['Meta Description'];
      const imageFilename = row['Image URL'];
      
      // Extract text from HTML
      const excerpt = truncate(extractTextFromHtml(excerptRaw), 500);
      const metaDescription = truncate(extractTextFromHtml(metaDescriptionRaw), 160);
      
      // Build image URL
      const coverImage = imageFilename ? `/images/blog/${imageFilename}` : null;
      
      // Check if article already exists
      const existing = await db.select().from(blogPosts).where(eq(blogPosts.slug, slug)).limit(1);
      if (existing.length > 0) {
        console.log(`⏭️  Skipping (exists): ${title.substring(0, 50)}...`);
        skipped++;
        continue;
      }
      
      // Insert article
      await db.insert(blogPosts).values({
        title,
        slug,
        content,
        excerpt,
        coverImage,
        category,
        tags,
        metaTitle,
        metaDescription,
        status: 'published',
        publishedAt: new Date(),
      });
      
      console.log(`✅ Imported: ${title.substring(0, 60)}...`);
      imported++;
      
    } catch (error) {
      console.error(`❌ Error importing article: ${error.message.substring(0, 100)}`);
      errors++;
    }
  }
  
  await connection.end();
  
  console.log(`\n${'='.repeat(60)}`);
  console.log('Import completed!');
  console.log(`✅ Imported: ${imported}`);
  console.log(`⏭️  Skipped: ${skipped}`);
  console.log(`❌ Errors: ${errors}`);
  console.log(`${'='.repeat(60)}`);
}

// Run import
importArticles().catch(console.error);
